base/cvd/cuttlefish/host/commands/cvd/server_command/start.cpp

```diff
@@ -52,6 +55,78 @@
 namespace cuttlefish {
 namespace {
 
+static constexpr int CLOSED_FD = -1;
+static constexpr int IN_USE_FD = -2;
+// Write end of the pipe for the signal handler. May hold the following values:
+// CLOSED_FD: Signals should not be sent through the pipe, if the thread that
+// owns the fd encounters this value it must close the fd. IN_USE_FD: A signal
+// was received and the handler is using the fd.
+// >= 0: The write end of the signal pipe.
+std::atomic<int> signal_pipe_write_end(CLOSED_FD);
+
+// Writes the signal number to the pipe if it's still open
+void InterruptHandler(int signal) {
+  auto fd = signal_pipe_write_end.exchange(IN_USE_FD);
+  if (fd >= 0) {
+    // Ignore result
+    write(fd, &signal, sizeof(signal));
+  }
+  fd = signal_pipe_write_end.exchange(fd);
+  if (fd != IN_USE_FD) {
+    // The signal handler was disabled while the handler was executing, need to close the fd.
+    fd = signal_pipe_write_end.exchange(CLOSED_FD);
+    if (fd >= 0) {
+      close(fd);
+    }
+  }
+}
+
+Result<void> InterruptProcessWaiterOnSignal(SubprocessWaiter* sw) {
+  int fds[2];
+  CF_EXPECTF(pipe2(fds, O_CLOEXEC) == 0, "Failed to create signals pipe: {}",
+             strerror(errno));
+
+  // Make the write end nonblocking
+  int flags = fcntl(fds[0], F_GETFL, 0);
+  fcntl(fds[0], F_SETFL, flags | O_NONBLOCK);
+
+  auto previous_value = signal_pipe_write_end.exchange(fds[0]);
+  CHECK(previous_value == CLOSED_FD) << "Interrupt handler set twice";
+
+  std::thread([sw, fd = fds[1]]() {
```

Comment:
Does `sw` live long enough for this (detached!) thread? A detached thread could live for the rest of the program lifetime.

Reply:
Fixed